### ISSUE TITLE
hevm: fails to deduct gas for memory expansion

### ISSUE THREAD

**Issue by wolflo** [dapphub/dapptools#543]
I haven't really gotten to the bottom of this, but in at least one case where the memory expansion caused by a `RETURN` op should cause an out of gas error, hevm succeeds without deducting any gas and returns the empty bytestring.

`return(0x0, max_uint)` using hevm and [geth](https://github.com/ethereum/go-ethereum)'s `evm` utility for comparison:
```
hevm exec --code 0x60016000036000f3 --gas 0xffffff
// VMSuccess: 0x

evm --code 0x60016000036000f3 --gas 0xffffff run
// error: gas uint64 overflow
```

hevm does run out of gas if the same memory access is made with non-empty memory:
`mstore(0x00, 0xff)`
`return(0x0, max_uint)`
```
hevm exec --code 0x60ff60005260016000036000f3 --gas 0xffffff
// OutOfGas 0xffffe7 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffd
```

Regardless of the memory contents, hevm throws `IllegalOverflow` if the accessed offset + length overflows 256 bits:
`return(0x01, max_uint)`
```
hevm exec --code 0x60016000036001f3 --gas 0xffffff
// IllegalOverflow
```

Edit:
```
hevm version // 0.42.0
```

**Comment by MrChico** [comment]
oh wow, thanks for reporting. The discussion in #342 is related, I think

**Comment by rainbreak** [comment]
Is there no GST that corresponds to this issue?

**Comment by MrChico** [comment]
The behavior displayed by hevm here comes from an overflow of uint256. I've adapted all gas calculations to be done in integers in the branch [gasoverflows](https://github.com/dapphub/dapptools/tree/gasoverflows), and set the limit to uint64.
geth uses uint64 for tracking gas, and the GST and geth and evmc have had a bit of a bad and forth on this in [3224](https://ethereum-magicians.org/t/eip-1985-sane-limits-for-certain-evm-parameters/3224), where the case is made to define uint64 to be desired behavior. I think this is reasonable.

The GSTs are not as exhaustive as one would like, I think much of the valuable testing is being done in fuzzing setups these days, such as ones engineered in github.com/ethereum/goevmlab. I think we can get some better milage out of this as well. I have implemented the type of json tracing discussed in #288 along with some `hypothesis` based cross reference testing in the branch `gethcompliance`, which lead me to discover some other intricacies that should be addressed.
